[GlobalISel] Add computeNumSignBits for G_SHUFFLE_VECTOR

davemgreen opened the pull request:
The code is similar to computeKnownBits and the code in SelectionDAG::ComputeNumSignBits.

arsenm commented on llvm/test/CodeGen/AArch64/aarch64-dup-ext.ll:
```diff
@@ -400,9 +400,10 @@ define <8 x i16> @missing_insert(<8 x i8> %b) {
-; CHECK-GI-NEXT:    sshll v0.8h, v0.8b, #0
-; CHECK-GI-NEXT:    ext v1.16b, v0.16b, v0.16b, #4
-; CHECK-GI-NEXT:    mul v0.8h, v1.8h, v0.8h
+; CHECK-GI-NEXT:    sshll v1.8h, v0.8b, #0
+; CHECK-GI-NEXT:    ext v1.16b, v1.16b, v1.16b, #4
+; CHECK-GI-NEXT:    xtn v1.8b, v1.8h
+; CHECK-GI-NEXT:    smull v0.8h, v1.8b, v0.8b
```
Regression

davemgreen replied:
Yeah - It looks like a missing trunc(shuffle(sext)) fold, you can see the SDAG code is still further ahead.

jayfoad commented on llvm/lib/CodeGen/GlobalISel/GISelValueTracking.cpp:
```diff
@@ -874,6 +874,30 @@ unsigned GISelValueTracking::computeNumSignBits(Register R,
+
+    unsigned Tmp = std::numeric_limits<unsigned>::max();
+    if (!!DemandedLHS)
+      Tmp =
+          computeNumSignBits(MI.getOperand(1).getReg(), DemandedLHS, Depth + 1);
```
Should early-out here if Tmp is 1, to save a second expensive call to computeNumSignBits.

jayfoad commented on llvm/lib/CodeGen/GlobalISel/GISelValueTracking.cpp:
```diff
@@ -874,6 +874,30 @@ unsigned GISelValueTracking::computeNumSignBits(Register R,
+    // If we don't know anything, early out and try computeKnownBits fall-back.
+    if (Tmp == 1)
+      break;
+    assert(Tmp <= TyBits && "Failed to determine minimum sign bits");
```
I don't think this is particularly useful. You could just initialize `Tmp` to `TyBits` instead.

arsenm commented on llvm/lib/CodeGen/GlobalISel/GISelValueTracking.cpp:
```diff
@@ -836,6 +837,28 @@ unsigned GISelValueTracking::computeNumSignBits(Register R,
+    // Collect the minimum number of sign bits that are shared by every vector
+    // element referenced by the shuffle.
+    APInt DemandedLHS, DemandedRHS;
+    unsigned NumElts = MRI.getType(MI.getOperand(1).getReg()).getNumElements();
```
Use temp register for MI.getOperand(1).getReg() and avoid the line break below

davemgreen committed:
[GlobalISel] Add computeNumSignBits for G_SHUFFLE_VECTOR

The code is similar to computeKnownBits and the code in
SelectionDAG::ComputeNumSignBits
